doc/neps/nep-0040-legacy-datatype-impl.rst

```diff
@@ -0,0 +1,616 @@
+``PyArray_Descr`` is a public structure.
+This is especially also true for the ``ArrFunctions`` structure stored in
+the ``f`` field.
+Due to compatibility they may need to remain supported for a very long time,
+with the possibility of replacing them by functions that dispatch to a newer API.
+
+However, in the long run access to these structures will probably have to
+be deprecated.
+
+
+NumPy Scalars and Type Hierarchy
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+As a side note to the above datatype implementation: unlike the datatypes,
+the NumPy scalars currently **do** provide a type hierarchy, including abstract
+types such as ``np.inexact`` (see figure below).
+In fact, some control flow within NumPy currently uses
+``issubclass(a.dtype.type, np.inexact)``.
+
+.. figure:: _static/nep-0040_dtype-hierarchy.png
+
+   **Figure:** Hierarchy of type objects reproduced from the reference
+   documentation. Some aliases such as ``np.intp`` are excluded. Datetime
+   and timedelta are not shown.
+
+NumPy scalars try to mimic zero dimensional arrays with a fixed datatype.
+For the numerical (and unicode) datatypes, they are further limited to
+native byte order.
+
+
+Current Implementation of Casting
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+One of the main features which datatypes need to support is casting between one
+another using ``arr.astype(new_dtype, casting="unsafe")``, or during execution
+of ufuncs with different types (such as adding integer and floating point numbers).
+
+Casting tables determine whether casting is possible or not.
+However, these cannot handle the parametric dtypes such as strings.
+The logic for parametric datatypes is defined mainly in ``PyArray_CanCastTo``
+and currently cannot be customized for user defined datatypes.
+
+The actual casting has two distinct parts:
+
+1. ``copyswap``/``copyswapn`` are defined for each dtype and can handle
+   byte-swapping for non-native byte orders as well as unaligned memory.
+2. The generic casting code is provided by C functions which know how to
+   cast aligned and contiguous memory from one dtype to another
+   (both in native byte order).
+3. C-level functions can be registered to cast aligned and contiguous memory
+   from one dtype to another.
+   The function may be provided with both arrays (although the parameter
+   is sometimes ``NULL`` for scalars).
+   NumPy will ensure that these functions receive native byte order input.
+   The current implementation stores the functions either in a C-array
+   on the datatype which is cast, or in a dictionary when casting to a user
+   defined datatype.
+
+When casting (small) buffers will be used when necessary to ensure
+contiguity, alignment or native byte order.
+In this case first ``copyswapn`` is called to and ensures that the cast function
+can handle the input.
+Generally NumPy will thus perform casting as chain of the three functions
+``in_copyswapn -> castfunc -> out_copyswapn``.
```

Comment:
This sentence is a bit redundant given the numbered points above, though I actually think this is more clear. Consider consolidating this section a bit.